staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go

```diff
@@ -419,10 +419,16 @@ func ObserveWaitingDuration(ctx context.Context, priorityLevel, flowSchema, exec
 // ObserveExecutionDuration observes the execution duration for flow control
 func ObserveExecutionDuration(ctx context.Context, priorityLevel, flowSchema string, executionTime time.Duration) {
 	reqType := "regular"
-	if requestInfo, ok := apirequest.RequestInfoFrom(ctx); ok && requestInfo.Verb == "watch" {
-		reqType = requestInfo.Verb
+	var verb, resource, subresource string
+	if requestInfo, ok := apirequest.RequestInfoFrom(ctx); ok {
+		if requestInfo.Verb == "watch" {
+			reqType = requestInfo.Verb
+		}
+		verb = requestInfo.Verb
+		resource = requestInfo.Resource
+		subresource = requestInfo.Subresource
```

Comment:
If we add these, we need the same guards put in place to prevent unbounded values that we have on other metrics

Reply:
[I'm not suggesting to ignore this problem - just asking question]
There are still metrics where we don't do anything, right? the core requestCounter metrics is one of those where IIUC we just put everything as is, right?
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L436

Why is that one non-problematic? @dgrisonnet

Reply:
As far as I am aware, the only metrics that have these labels are the ones used to compute the kube-apiserver SLOs. So `kube_apiserver_request_total` and `kube_apiserver_request_duration`.

These ones are actually problematic, we are talking about metrics that represent around 50% of the timeseries exposed by the apiserver. Over the years, there have been many complaints about these metrics resulting in too many series, but most of them were discarded as these metrics are essential to know the apiserver's availability.

So I would say that these metrics are problematic, but because of their importance, we lived with the memory impact that they could bring and I don't think any other metrics are important enough to justify this cost.

That said, we might want to consider changing that. Bounding the number of resources values exposed doesn't sound like a bad idea. Even more since we have added latency information to the audit logs (https://github.com/kubernetes/kubernetes/pull/107910). With that, even if we get rid of most of the resource-level information, we would still be able to get it from the audit logs.

Going even further, if one was to hide all the resource-level information today (e.g dropping the label during ingestion), it wouldn't impact the SLO computation since it isn't required there, and they would still be able to investigate issues at a resource-level from the logs. So I think that we might be able to make nice memory gains by introducing a resource allowlist for these metrics.

Reply:
> So I would say that these metrics are problematic, but because of their importance, we lived with the memory impact that they could bring and I don't think any other metrics are important enough to justify this cost.

OK - that makes sense to me.

> Going even further with that idea, if one was to hide all the resource-level information today (e.g dropping the labels at ingestion), it wouldn't impact the SLO computation since it isn't required there, and they would still be able to investigate issues at a resource-level from the logs. So I think that we might be able to make nice memory gains by introducing a resource allowlist for these metrics.

I disagree with this one. We're defining SLO per (resource, verb) pair: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md so unless I misunderstood something I don't think we will be able to compute it. And I don't think we should be going towards computing SLOs from audit logs.
Also note - that audit logs don't contain all the requests - IIRC even the default policies remove some system-originating (read) requests, which we really want to have in SLOs. We may try doing a bit of bundling, but I don't agree we can actually drop the labels.

So as long as I agree we shouldn't be adding more metrics with such high cardinality, I don't really agree we are ready to drop labels from the two you mentioned above.

Reply:
I was slightly wrong, I forgot that we are using the subresource label for SLOs: https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go#L48. But, I don't think we are using the resource label at all there. We are not defining SLO threshold based on each resource latency, but rather based on their scope since that is where we can see big differences. The subresource filter is just here to handle a few corner cases.

I can see how useful these labels are today, but I can also see a potential world without any values for them, or only a few allow listed ones in order to reduce costs.

> Also note - that audit logs don't contain all the requests - IIRC even the default policies remove some system-originating (read) requests, which we really want to have in SLOs. We may try doing a bit of bundling, but I don't agree we can actually drop the labels.

We won't be able to drop the labels from the codebase anyway since the metrics are stable, but we might still be able to filter them.
I wasn't aware that some requests are not present in the audit logs. As I mentioned before, I don't think how SLOs are computed will be impacted, but if we start filtering some resources, we might be dropping the only information we have about these requests which might be problematic.

Reply:
> But, I don't think we are using the resource label at all there. We are not defining SLO threshold based on each resource latency, but rather based on their scope since that is where we can see big differences. The subresource filter is just here to handle a few corner cases.

No - the SLO is per (resource, verb) pair. So each (resource, verb) has to be satisfied on your own (see the definition I linked above).